elyra/pipeline/registry.py

```diff
@@ -0,0 +1,483 @@
+#
+# Copyright 2018-2020 IBM Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import os
+import io
+import json
+import yaml
+import urllib
+
+from traitlets.config import SingletonConfigurable
+
+
+cardinality = {
+    'min': 0,
+    'max': -1
+}
+
+inputs = {
+    "id": "inPort",
+    "app_data": {
+        "ui_data": {
+            "cardinality": cardinality,
+            "label": "Input Port"
+        }
+    }
+}
+
+outputs = {
+    "id": "outPort",
+    "app_data": {
+        "ui_data": {
+            "cardinality": cardinality,
+            "label": "Output Port"
+        }
+    }
+}
+
+default_current_parameters = {
+    "filename": "",
+    "runtime_image": "",
+    "outputs": [],
+    "env_vars": [],
+    "dependencies": [],
+    "include_subdirectories": False
+}
+
+
+def get_id_from_name(name):
+    return ' '.join(name.lower().replace('-', '').split()).replace(' ', '-').replace('_', '-')
+
+
+def set_node_type_data(id, label, description):
+    node_type = {
+        'id': "",
+        'op': f"execute-{id}-node",
+        'type': "execution_node",
+        'inputs': [inputs],
+        'outputs': [outputs],
+        'parameters': {},
+        'app_data': {
+            'ui-data': {
+                'label': label,
+                'description': description,
+                'image': "",
+                'x_pos': 0,
+                'y_pos': 0
+            }
+        }
+    }
+
+    return node_type
+
+
+class ComponentParser(SingletonConfigurable):
+    _type = "local"
+    properties: dict() = {}
+    parameters: dict() = {}
+
+    def __init__(self):
+        super().__init__()
+        self.properties = self.get_common_config('properties')
+
+    def get_common_config(self, config_name):
+        common_dir = os.path.join(os.path.dirname(__file__), 'resources')
+        common_file = os.path.join(common_dir, f"{config_name}.json")
+        with io.open(common_file, 'r', encoding='utf-8') as f:
+            common_json = json.load(f)
+
+        return common_json
+
+    def _get_component_catalog_json(self):
+        catalog_dir = os.path.join(os.path.dirname(__file__), "resources")
+        catalog_file = os.path.join(catalog_dir, f"{self._type}_component_catalog.json")
+        with open(catalog_file, 'r') as f:
+            catalog_json = json.load(f)
+
+        return catalog_json['components']
```

Comment:
Rather than a List, should this return a Dict indexed by the component's id?  This way, `list_all_components()` becomes: `return self._get_component_catalog_json().values()` and `return_component_if_exists(id)` becomes `return self._get_component_catalog_json().get(id)`.

Reply:
That makes a lot of sense!